### PR TITLE
Change old $DELETE to $REMOVE, and new implement $DELETE to follow AWS DynamoDB

### PR DESCRIPTION
**Pull request by CodeBanBan**
### Summary:
**BREAKING CHANGE:** Change old $DELETE to $REMOVE, and new implement $DELETE to follow AWS DynamoDB

### Is this ready to be merged into Dynamoose? (select 1):
- [ ] Yes
- [ ] No
- [ x ] Not sure

### Other:
- [ x ] I have searched through the GitHub pull requests to ensure this PR has not already been submitted
- [ ] I have updated the Dynamoose documentation (if required) given the changes I made
- [ x ] I have added/updated the Dynamoose test cases (if required) given the changes I made
- [ ] I agree that all changes made in this pull request may be distributed and are made available in accordance with the [Dynamoose license](https://github.com/dynamoosejs/dynamoose/blob/master/LICENSE.txt)
- [ x ] All of my commits and commit messages are detailed, explain what changes were made, and are easy to follow and understand
- [ ] I have filled out all fields above

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,4 +1,10 @@
-# Dynamoose
+# Dynamoose-mod
+
+This Fork from [Dynamoose](https://www.npmjs.com/package/dynamoose)
+### Update Feature
+- Update Model.update follow to AWS DynamoDB Behavior
+    - Refactor Model.update $DELETE to $REMOVE
+    - Implement Model.update support $DELETE to delete item to set
```
Comment (fishcharlie) [inline README.md]: This should all be removed since it’s not a fork anymore.

File: lib/Model.js [hunk 2]
```diff
@@ -991,6 +1016,25 @@ Model.update = async function (NewModel, key, update, options, next) {
     }
   }
 
+  if (update.$DELETE) {
+    for (const deleteItem in update.$DELETE) {
+      const deleteAttr = schema.attributes[deleteItem];
+      const deleteVal = update.$DELETE[deleteItem];
+      try {
+        if (deleteAttr === 'list') {
+          const err = new Error('$DELETE not support type list');
```
Comment (fishcharlie) [inline lib/Model.js]:
```suggestion
          const err = new Error('$DELETE does not support type list');
```

File: package.json [hunk 3]
```diff
@@ -1,15 +1,12 @@
 {
-  "name": "dynamoose",
+  "name": "dynamoose-mod",
```
Comment (fishcharlie) [inline package.json]: This should be reversed.

File: package.json [hunk 4]
```diff
@@ -109,8 +106,7 @@
   "husky": {
     "hooks": {
       "commit-msg": "commitlint -E HUSKY_GIT_PARAMS",
-      "pre-commit": "npm run lint && npm run tslint",
-      "pre-push": "npm run ci"
+      "pre-commit": "npm run lint && npm run tslint"
```
Comment (fishcharlie) [inline package.json]: This should be reversed.

File: package.json [hunk 5]
```diff
@@ -1,15 +1,12 @@
...
   "description": "Dynamoose is a modeling tool for Amazon's DynamoDB (inspired by Mongoose)",
   "typings": "./dynamoose.d.ts",
   "homepage": "https://dynamoosejs.com",
   "repository": {
     "type": "git",
-    "url": "https://github.com/dynamoosejs/dynamoose"
-  },
-  "bugs": {
-    "url": "https://github.com/dynamoosejs/dynamoose/issues"
+    "url": "https://github.com/CodeBanBan/dynamoose"
```
Comment (fishcharlie) [inline package.json]: This should be reversed.

**Comment by fishcharlie** [comment]
@CodeBanBan Also looks like there are some merge conflicts to fix.

@hweeks I’m pretty sure our deployment tools will mark this as a breaking change, and bump the version to 2.0.0, but could you please verify I’m accurate about that? I don’t wanna release this and not have it be a breaking change.

**Comment by dolsem** [comment]
I am reviewing it now to confirm this, but seems like the goal of this PR is to update `$DELETE` and `$REMOVE` to match `DELETE` and `REMOVE` from DynamoDB. After this we will have `$PUT`, `$ADD`, `$REMOVE` and `$DELETE`. I don't remember if `$ADD` precisely matches `ADD` from DynamoDB, but I think it does. However, DynamoDB does not have a `PUT` operation, but it does have a `SET` operation. So if this PR is merged and released as version 2.0.0, it still won't match Update Expressions from DynamoDB precisely.

**Comment by fishcharlie** [comment]
@dolsem So what else does this PR have to do to match DynamoDB as much as possible? I'm not QUITE as knowledgeable about this area of DynamoDB.

File: test/Model.js [hunk 6]
```diff
@@ -2298,8 +2298,27 @@ describe('Model', function () {
         });
       });
     });
+
+    it('Delete item to list', (done) => {
```
Comment (dolsem) [inline test/Model.js]:
```suggestion
    it('Delete item from list', (done) => {
```

**Comment by hweeks** [comment]
This will release a v2.0 for sure @fishcharlie. Let's make sure to get that done and also get parity per @dolsem's noted fixes/updates.

**Comment by fishcharlie** [comment]
@hweeks or @dolsem If either of you have time and would like to make an alternate PR to try to get this in faster feel free. I feel bad about snoozing on this issue for so long so trying to do as much as possible to get this in as fast as possible haha.

**Comment by dolsem** [comment]
Here's an overview of [Update Expressions operations from DynamoDB docs](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/Expressions.UpdateExpressions.html):
### SET
- If supplied with an attribute that _**does not exist**_ on the item, _**adds**_ that attribute to the item. 
- If supplied with an attribute that _**exists**_ on the item, _**overwrites**_ that attribute with the new value.
- Can be used to overwrite elements of a list given index, or append to the end of the list if you provide an index value equal to the length of the list. E.g. `` `SET List[1] = :item` ``.
- Supports nested paths, for example given document:
```js
{
  id: 1, // Number type, N
  map: { // Map type, M
    list: ['item0'] // List type, L
  }
}
```
`` `SET map.list[1] = :item1` `` will append to the list, and `` `SET map.number = :number` `` will add a new entry to the map with key `number` and value of `:number`.
- Expressions can be used instead of literal values. These expressions can include functions and references to other fields or nested values.
- Can be used to increment/decrement numbers, for example `` `SET Price = Price - :number` ``.
- Supports `list_append` function which takes two lists as its input (so technically works more like concat, not append): `` `SET priceList = list_append(:new_prices, priceList)` `` will add all prices in `:new_prices` to the beginning of `priceList`.
- Can set different values depending whether an attribute exists or not: `` `SET price = if_not_exists(newPrice, oldPrice)` `` will set `price` to `newPrice` if `newPrice` exists and to `oldPrice` otherwise.
### REMOVE
- If attribute exists will remove that attribute.
- If attribute does not exist nothing happens, but query still succeeds.
- Can remove element from a list given an index: `` `REMOVE list[1], list[4], list[7]` ``.
### ADD
- Used _**exclusively**_ to update numbers and add elements to sets.
- Only supports literals, does not support expressions (unlike SET).
- If provided attribute is a number it will add provided value to it (or subtract, if provided value is negative): `` `ADD velocity :accelation_times_duration` ``.
- Given `":new_items": {"SS":["itemFour", "itemFive"]}` and document:
```js
{
  id: 1, // Number type, N
  setA: ['itemOne', 'itemTwo', 'itemThree'], // Set type, SS
}
```
`` `ADD setA :new_items` `` will result in `setA: ['itemOne', 'itemTwo', 'itemThree', 'itemFour', 'itemFive']`. `` `ADD setB :new_items` `` will result in attribute `setB` being added to the document - `setB: ['itemFour', 'itemFive']`
- It is not possible to do `` `ADD setA :new_item` `` to add a single item to the set, provided value has to be of set type.
- DynamoDB docs recommend to use SET for number operations instead of ADD
### DELETE
- Used _**exclusively**_ to remove elements from sets.
- `` `DELETE setA :items` `` will delete specified items from setA. `:items` must be of Set type.

**Comment by dolsem** [comment]
@fishcharlie Personally I need to deploy a project I am working on to production asap, and I have to finish optimistic locking before I can do that, so I don't think I'll be able to contribute to this more at least until I finish the plugin.

Hopefully, my last comment was helpful in clearing up some confusion we all had about how Update Expressions operate. I also hope that this information will end up in the docs, it would be helpful to summarize it to newcomers and provide examples, so that they don't have to go through official DynamoDB docs. 

If we want `Model.update()` to be 100% aligned with DynamoDB I think the first step should be to write comprehensive tests for all of this functionality I laid out. Hopefully @CodeBanBan will have the time to work out all the remaining pieces. If not, we should probably merge this PR to a different branch for now, and wait for the rest of the code before updating to 2.0.0.

File: BREAKING_CHANGES.md [hunk 7]
```diff
@@ -0,0 +1,8 @@
+# Breaking Change
+
```
Comment (fishcharlie) [inline BREAKING_CHANGES.md]: I think adding the following will be a good idea as we do more breaking changes in the future.

```suggestion

## v1.x to v2.x
```

**Comment by hweeks** [comment]
that’s not needed charlie. the release notes will capture it

**Comment by fishcharlie** [comment]
@hweeks The goal is to be as explicit as possible and give more detail about how to upgrade. Will the release notes really give as much detail as that migration guide does?

Just trying to figure out how to make it as painless as possible for people to find the information to upgrade and fix their code.

**Comment by CodeBanBan** [comment]
@fishcharlie For idea try to have complete DynamoDB parity from @dolsem
I think I need more time because I quite busy in next 1-2 month,
If we can wait, I think I can do it
PS. I'm sorry for my english not well

**Comment by CodeBanBan** [comment]
@fishcharlie I review code and I think just change $PUT to $SET
other operation ($ADD, $REMOVE, $DELETE) work ok follow dynamodb
I will update to $SET as soon as possible

**Comment by fishcharlie** [comment]
@CodeBanBan Let me know when you are ready for me to review & merge this.

**Comment by CodeBanBan** [comment]
@fishcharlie Sorry for late, now I ready for your review

**Comment by fishcharlie** [comment]
@CodeBanBan I think we will probably handle this as part of the v2.0 rewrite (#677).

**Comment by john-aws** [comment]
Per @dolsem's earlier [comment](https://github.com/dynamoose/dynamoose/pull/644#issuecomment-506465236) on standard DynamoDB update expressions, are there any plans to implement support for standard, pretty important DynamoDB update expressions that involve lists or attribute references, such as:

```
ADD #listname[1] 24
SET mycount = mycount + 1
SET #value = #value + 1
SET #listname[1] = #listname[0] + 15
```

None of these appear to be natively supported by Dynamoose at present, and there appears to be no way to work around this other than to use the AWS SDK directly.

**Pull request closed without merging**